app/routes/profile.tsx

```diff
@@ -0,0 +1,92 @@
+import { json, redirect, Session } from "@remix-run/node";
+import type { LoaderFunction } from "@remix-run/node";
+import type { FC } from "react";
+import { commitSession, destroySession, getSession } from "~/session.server";
+import { useLoaderData } from "@remix-run/react";
+import RootLayout from "~/components/RootLayout";
+import { auth } from "~/firebase.server";
+
+/**
+ * Returns the user that's currently logged in, or null if there is none.
+ *
+ * @see https://firebase.google.com/docs/auth/admin/verify-id-tokens#web
+ */
+export async function getCurrentUser(session: Session) {
+  const idToken = session.get("idToken");
+
+  const error = session.get("error");
+  if (error) {
+    throw new Error(error);
+  }
+
+  if (typeof idToken !== "string") {
+    return null;
+  }
+
+  try {
+    const decodedClaims = await auth.verifyIdToken(idToken);
+    return decodedClaims;
+  } catch (err) {
+    // The session isn't valid anymore so return undefined
+    return null;
+  }
+}
+
+export const loader: LoaderFunction = async ({ request }) => {
```

Comment:
In Remix, routes can export a `loader()` method. This runs on the server and can be used to load data.

In this case, we check for a session cookie, make sure it contains an `idToken`, and verify that token using the `firebase-admin` SDK. If all goes well, we'll have decoded claims that contain information about the user and we send that down to the component. If anything goes wrong, we clear the session and redirect to the `/login` age.